src/tunnel.cc

```diff
@@ -940,15 +925,17 @@ tunnelStartShoveling(TunnelStateData *tunnelState)
  * Call the tunnelStartShoveling to start the blind pump.
  */
 static void
-tunnelConnectedWriteDone(const Comm::ConnectionPointer &conn, char *, size_t len, Comm::Flag flag, int, void *data)
+tunnelConnectedWriteDone(const Comm::ConnectionPointer &conn, char *, size_t len, Comm::Flag flag, int xerrno, void *data)
 {
     TunnelStateData *tunnelState = (TunnelStateData *)data;
     debugs(26, 3, HERE << conn << ", flag=" << flag);
     tunnelState->client.writer = NULL;
 
     if (flag != Comm::OK) {
-        *tunnelState->status_ptr = Http::scInternalServerError;
-        tunnelErrorComplete(conn->fd, data, 0);
+        const auto err = new ErrorState(ERR_WRITE_ERROR, Http::scInternalServerError, tunnelState->request.getRaw(), tunnelState->al);
+        err->xerrno = xerrno;
+        tunnelState->saveError(err);
+        tunnelState->sendErrorAndDestroy("CONNECT response write failure");
```

Comment:
Please add a comment to the TunnelStateData::Connection::error() method. Something along the lines of "No sendErrorAndDestroy(): We are only called when errors cannot be sent". I cannot attach this ([old](https://github.com/measurement-factory/squid/pull/112#pullrequestreview-603133234)) change request to that code.